### PR TITLE
[ML] Adds Explain Functionality to LTR Rescoring

### PR DESCRIPTION
**Pull request by markjhoy**
Adds explain functionality to LTR rescoring.

The explanations will be output (using the `explain=true` flag in the query) and will be per result hit describing the query and the features used in the rescoring as well as the weights that contributed to the rescoring:
```
                "_explanation": {
                    "value": 20,
                    "description": "rescored using LTR model ltr-model",
                    "details": [
                        {
                            "value": 1,
                            "description": "first pass query score",
                            "details": [
                                {
                                    "value": 1,
                                    "description": "*:*",
                                    "details": []
                                }
                            ]
                        },
                        {
                            "value": 0,
                            "description": "extracted features",
                            "details": [
                                {
                                    "value": 1,
                                    "description": "feature value for [type_tv]",
                                    "details": []
                                },
                                {
                                    "value": 300,
                                    "description": "feature value for [cost]",
                                    "details": []
                                },
                                {
                                    "value": 2,
                                    "description": "feature value for [two]",
                                    "details": []
                                }
                            ]
                        }
                    ]
                }
```

### CODE REVIEW

**Comment by markjhoy** [comment]
Also - I'm trying to add the team labels, but keep getting:
[elasticsearchmachine](https://github.com/elasticsearchmachine) removed [Team:ML](https://github.com/elastic/elasticsearch/labels/Team%3AML) [Team:Search Relevance](https://github.com/elastic/elasticsearch/labels/Team%3ASearch%20Relevance) labels [now](https://github.com/elastic/elasticsearch/pull/112155#event-14014636373)

**Comment by afoucret** [comment]
The label is added automatically based on the are of your PR. I added `:Search Relevance/Ranking` as area and `elasticmachine` added the right team automatically.

**Comment by afoucret** [comment]
I will review your PR tomorrow

File: x-pack/plugin/ml/qa/single-node-tests/src/javaRestTest/java/org/elasticsearch/xpack/ml/integration/LearningToRankRescorerIT.java [hunk 1]
```diff
@@ -28,189 +30,11 @@ public class LearningToRankRescorerIT extends InferenceTestCase {
...
-        indexData("{ \"product\": \"TV\", \"cost\": 300}");
-        indexData("{ \"product\": \"TV\", \"cost\": 400}");
-        indexData("{ \"product\": \"TV\", \"cost\": 600}");
-        indexData("{ \"product\": \"VCR\", \"cost\": 15}");
-        indexData("{ \"product\": \"VCR\", \"cost\": 350}");
-        indexData("{ \"product\": \"VCR\", \"cost\": 580}");
-        indexData("{ \"product\": \"Laptop\", \"cost\": 100}");
-        indexData("{ \"product\": \"Laptop\", \"cost\": 300}");
-        indexData("{ \"product\": \"Laptop\", \"cost\": 500}");
+        putRegressionModel(MODEL_ID, testRegressionModel);
```
Comment (kderusso) [inline x-pack/plugin/ml/qa/single-node-tests/src/javaRestTest/java/org/elasticsearch/xpack/ml/integration/LearningToRankRescorerIT.java:33]: Nice cleanup 👍

**Review by kderusso — COMMENTED** [review]
Changes look reasonable to me

**Review by afoucret — APPROVED** [review]

**Comment by afoucret** [comment]
I will merge the PR since @markjhoy is out for the next few days.

Great work!!!

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/ltr/LearningToRankRescorer.java [hunk 2]
```diff
@@ -129,8 +130,58 @@ public TopDocs rescore(TopDocs topDocs, IndexSearcher searcher, RescoreContext r
     @Override
     public Explanation explain(int topLevelDocId, IndexSearcher searcher, RescoreContext rescoreContext, Explanation sourceExplanation)
         throws IOException {
-        // TODO: Call infer again but with individual feature importance values and explaining the model (which features are used, etc.)
-        return null;
+        if (sourceExplanation == null) {
+            return Explanation.noMatch("no match found");
+        }
```
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/ltr/LearningToRankRescorer.java:135]: Is there a reason other than expediency for why we aren't returning shapley values?

The stored models should be able to return the individual feature importance for the individual doc inference, not just what feature values were.

**Pull request merged**